File: scripts/audit-translations.js
#!/usr/bin/env node
/**
 * WorldForge Translation Audit Script
 *
 * Valideert dat alle wft(...) calls in JS files corresponderende keys hebben
 * in zowel nl.json als en.json.
 *
 * Gebruik: node scripts/audit-translations.js
 *
 * Exit code: 0 (OK) of 1 (fouten gevonden)
 */

const fs = require('fs');
const path = require('path');
const glob = require('glob');

// ANSI colors voor output
const colors = {
  reset: '\x1b[0m',
  red: '\x1b[31m',
  green: '\x1b[32m',
  yellow: '\x1b[33m',
  blue: '\x1b[36m'
};

const log = {
  error: (msg) => console.log(`${colors.red}✗${colors.reset} ${msg}`),
  success: (msg) => console.log(`${colors.green}✓${colors.reset} ${msg}`),
  warn: (msg) => console.log(`${colors.yellow}⚠${colors.reset} ${msg}`),
  info: (msg) => console.log(`${colors.blue}ℹ${colors.reset} ${msg}`)
};

// Laad translation files
function loadTranslations(lang) {
  const filePath = path.join(__dirname, `../lang/${lang}.json`);
  try {
    const content = fs.readFileSync(filePath, 'utf-8');
    return JSON.parse(content);
  } catch (err) {
    log.error(`Kan ${lang}.json niet laden: ${err.message}`);
    process.exit(1);
  }
}

// Vind alle wft() calls in JS files
function findTranslationKeys() {
  const jsFiles = glob.sync(path.join(__dirname, '**/*.js'), {
    ignore: [
      path.join(__dirname, 'audit-translations.js'),
      path.join(__dirname, 'node_modules/**')
    ]
  });

  const keys = new Set();
  const keysWithFiles = new Map(); // key -> [files where used]

  jsFiles.forEach(file => {
    try {
      const content = fs.readFileSync(file, 'utf-8');
      // Regex: wft("KEY") of wft('KEY') of wft(\`KEY\`)
      const regex = /wft\s*\(\s*["'`]([A-Z0-9._]+)["'`]\s*\)/g;
      let match;

      while ((match = regex.exec(content)) !== null) {
        const key = match[1];
        keys.add(key);

        if (!keysWithFiles.has(key)) {
          keysWithFiles.set(key, []);
        }
        keysWithFiles.get(key).push(path.relative(process.cwd(), file));
      }
    } catch (err) {
      log.warn(`Kan ${file} niet lezen: ${err.message}`);
    }
  });

  return { keys: Array.from(keys).sort(), keysWithFiles };
}

// Main audit
function audit() {
  console.log('\n📋 WorldForge Translation Audit\n');

  const nlTranslations = loadTranslations('nl');
  const enTranslations = loadTranslations('en');
  const { keys, keysWithFiles } = findTranslationKeys();

  log.info(`Gevonden ${keys.length} unieke translation keys in code`);
  log.info(`NL translations: ${Object.keys(nlTranslations).length} keys`);
  log.info(`EN translations: ${Object.keys(enTranslations).length} keys`);

  console.log('\n--- VALIDATIE ---\n');

  let errors = 0;
  const missing = {
    nl: [],
    en: []
  };

  keys.forEach(key => {
    const inNl = key in nlTranslations;
    const inEn = key in enTranslations;

    if (!inNl) {
      missing.nl.push(key);
      errors++;
    }
    if (!inEn) {
      missing.en.push(key);
      errors++;
    }

    if (inNl && inEn) {
      // OK
    } else {
      const status = [];
      if (!inNl) status.push('NL ✗');
      if (!inEn) status.push('EN ✗');
      log.error(`${key} — ${status.join(', ')}`);

      const files = keysWithFiles.get(key);
      files.forEach(f => console.log(`     └─ ${f}`));
    }
  });

  console.log('\n--- SAMENVATTING ---\n');

  if (errors === 0) {
    log.success('Alle translation keys zijn compleet!');
    console.log();
    return 0;
  } else {
    log.error(`${errors} ontbrekende translations gevonden\n`);

    if (missing.nl.length > 0) {
      console.log(`${colors.yellow}Ontbreekt in nl.json:${colors.reset}`);
      missing.nl.forEach(key => {
        console.log(`  "${key}": ""`);
      });
    }

    if (missing.en.length > 0) {
      console.log(`\n${colors.yellow}Ontbreekt in en.json:${colors.reset}`);
      missing.en.forEach(key => {
        console.log(`  "${key}": ""`);
      });
    }

    console.log();
    return 1;
  }
}

// Run
const exitCode = audit();
process.exit(exitCode);

File: scripts/audit-translations.py
#!/usr/bin/env python3
"""
WorldForge Translation Audit Script

Valideert dat alle wft(...) calls in JS files corresponderende keys hebben
in zowel nl.json als en.json.

Gebruik: python scripts/audit-translations.py
"""

import json
import re
import os
import sys
from pathlib import Path

# Force UTF-8 output on Windows
if sys.platform == 'win32':
    import io
    sys.stdout = io.TextIOWrapper(sys.stdout.buffer, encoding='utf-8')

# ANSI colors
class Colors:
    RED = '\033[91m'
    GREEN = '\033[92m'
    YELLOW = '\033[93m'
    BLUE = '\033[96m'
    RESET = '\033[0m'

def log_error(msg):
    print(f"{Colors.RED}✗{Colors.RESET} {msg}")

def log_success(msg):
    print(f"{Colors.GREEN}✓{Colors.RESET} {msg}")

def log_warn(msg):
    print(f"{Colors.YELLOW}⚠{Colors.RESET} {msg}")

def log_info(msg):
    print(f"{Colors.BLUE}ℹ{Colors.RESET} {msg}")

def load_translations(lang):
    """Laad translation JSON file"""
    file_path = Path(__file__).parent.parent / f"lang/{lang}.json"
    try:
        with open(file_path, 'r', encoding='utf-8') as f:
            return json.load(f)
    except Exception as err:
        log_error(f"Kan {lang}.json niet laden: {err}")
        sys.exit(1)

def find_translation_keys():
    """Vind alle wft() calls in JS files"""
    project_dir = Path(__file__).parent.parent
    keys = set()
    keys_with_files = {}

    # Vind alle JS files recursief (behalve audit-translations.js)
    js_files_found = list(project_dir.rglob("*.js"))

    for js_file in js_files_found:
        if "audit-translations" in js_file.name or "node_modules" in str(js_file):
            continue

        try:
            with open(js_file, 'r', encoding='utf-8') as f:
                content = f.read()

            # Regex: wft("KEY") of wft('KEY') - aanpassen voor alle cases
            # Matches: WF.*, WORLDFORGE.*, enz.
            regex = r'wft\s*\(\s*["\']([A-Za-z0-9._]+)["\']\s*\)'
            matches = re.findall(regex, content)

            for key in matches:
                keys.add(key)
                if key not in keys_with_files:
                    keys_with_files[key] = []
                rel_path = js_file.relative_to(project_dir)
                keys_with_files[key].append(str(rel_path))
        except Exception as err:
            log_warn(f"Kan {js_file.name} niet lezen: {err}")

    return sorted(list(keys)), keys_with_files

def audit():
    """Voer audit uit"""
    print("\n📋 WorldForge Translation Audit\n")

    nl_translations = load_translations('nl')
    en_translations = load_translations('en')
    keys, keys_with_files = find_translation_keys()

    log_info(f"Gevonden {len(keys)} unieke translation keys in code")
    log_info(f"NL translations: {len(nl_translations)} keys")
    log_info(f"EN translations: {len(en_translations)} keys")

    print("\n--- VALIDATIE ---\n")

    errors = 0
    missing_nl = []
    missing_en = []

    for key in keys:
        in_nl = key in nl_translations
        in_en = key in en_translations

        if not in_nl:
            missing_nl.append(key)
            errors += 1
        if not in_en:
            missing_en.append(key)
            errors += 1

        if in_nl and in_en:
            continue  # OK
        else:
            status = []
            if not in_nl:
                status.append("NL ✗")
            if not in_en:
                status.append("EN ✗")
            log_error(f"{key} — {', '.join(status)}")

            files = keys_with_files.get(key, [])
            for f in files:
                print(f"     └─ {f}")

    print("\n--- SAMENVATTING ---\n")

    if errors == 0:
        log_success("Alle translation keys zijn compleet!")
        print()
        return 0
    else:
        log_error(f"{errors} ontbrekende translations gevonden\n")

        if missing_nl:
            print(f"{Colors.YELLOW}Ontbreekt in nl.json:{Colors.RESET}")
            for key in missing_nl:
                print(f'  "{key}": ""')

        if missing_en:
            print(f"\n{Colors.YELLOW}Ontbreekt in en.json:{Colors.RESET}")
            for key in missing_en:
                print(f'  "{key}": ""')

        print()
        return 1

if __name__ == "__main__":
    exit_code = audit()
    sys.exit(exit_code)
